constraint/pkg/client/drivers/local/instrumentation.go

```diff
@@ -0,0 +1,80 @@
+package local
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/open-policy-agent/frameworks/constraint/pkg/client/drivers"
+	"github.com/open-policy-agent/frameworks/constraint/pkg/instrumentation"
+)
+
+// TODO docs, not thread safe.
+type metricsReporterImpl struct {
+	store map[drivers.ConstraintKey][]instrumentation.Stats
+	// TODO not the best way to show underlying issues in reporting
+	errorsWhileObserving bool
+}
+
+func (m *metricsReporterImpl) EvaluationStats() string {
+	var b strings.Builder
+	for ck, stats := range m.store {
+		for _, stat := range stats {
+			b.WriteString(fmt.Sprintf("Constraint Kind: %s, Constraint Name: %s, Evaluation Metadata: %s", ck.Kind, ck.Name, stat.ToString()))
+		}
+	}
+	if m.errorsWhileObserving {
+		b.WriteString("\nWARN: reporter encountered an error with the data format while observing engine evaluation; some stats may have been dropped.\n")
+	}
+
+	return b.String()
+}
+
+func (m *metricsReporterImpl) StatsEntries() []*instrumentation.StatsEntry {
+	if m.store == nil {
+		return nil
+	}
+
+	var entries []*instrumentation.StatsEntry
```

Comment:
nit:
`entries := make([]*instrumentation.StatsEntry, len(m.store))`
might be more efficient. If the slice isn't going to grow: `make([]*instrumentation.StatsEntry, len(m.store), len(m.store))

Reply:
n/a know after `4cdb922` bc we are not using a Reporter anymore